File: internal/model/stats.go
package model

type StatsMetrics struct {
	InputToken     int64   `json:"input_token" gorm:"bigint"`
	OutputToken    int64   `json:"output_token" gorm:"bigint"`
	InputCost      float64 `json:"input_cost" gorm:"type:real"`
	OutputCost     float64 `json:"output_cost" gorm:"type:real"`
	WaitTime       int64   `json:"wait_time" gorm:"bigint"`
	RequestSuccess int64   `json:"request_success" gorm:"bigint"`
	RequestFailed  int64   `json:"request_failed" gorm:"bigint"`
}

type StatsTotal struct {
	ID int `gorm:"primaryKey"`
	StatsMetrics
}

type StatsHourly struct {
	Hour int    `json:"hour" gorm:"primaryKey"`
	Date string `json:"date" gorm:"not null"` // 记录最后更新日期，格式：20060102
	StatsMetrics
}

type StatsDaily struct {
	Date string `json:"date" gorm:"primaryKey"`
	StatsMetrics
}

type StatsAPIKey struct {
	APIKeyID int `json:"api_key_id" gorm:"primaryKey"`
	StatsMetrics
}

// StatsChannelDaily 是单个渠道在某一天的统计。
// 渠道自身只存累计统计, 无从按时间切片, 故首页按周期切换榜单时需要这一份按日明细。
// 只留最近若干天: 榜单最长看 30 天, 全时段直接读渠道上的累计列, 无限保留只会让表随运行时长增长。
type StatsChannelDaily struct {
	ChannelID    int    `json:"channel_id" gorm:"primaryKey"` // 渠道主键。
	Date         string `json:"date" gorm:"primaryKey;index"` // 统计日期, 格式 20060102; 按日期清理与范围查询都走该索引。
	StatsMetrics        // 该渠道当日的统计。
}

// StatsChannelModelDaily 是单个渠道模型在某一天的统计, 与 StatsChannelDaily 同理。
type StatsChannelModelDaily struct {
	ChannelModelID int    `json:"channel_model_id" gorm:"primaryKey"` // 渠道模型主键。
	Date           string `json:"date" gorm:"primaryKey;index"`       // 统计日期, 格式 20060102。
	StatsMetrics          // 该渠道模型当日的统计。
}

// Add aggregates another StatsMetrics into the current one.
func (s *StatsMetrics) Add(delta StatsMetrics) {
	s.InputToken += delta.InputToken
	s.OutputToken += delta.OutputToken
	s.InputCost += delta.InputCost
	s.OutputCost += delta.OutputCost
	s.WaitTime += delta.WaitTime
	s.RequestSuccess += delta.RequestSuccess
	s.RequestFailed += delta.RequestFailed
}

File: web/src/api/stats.ts
import { queryOptions, useQuery } from '@tanstack/react-query';
import { apiRequest } from './client';
import { statsDailyQueryOptions, statsHourlyQueryOptions, statsTotalQueryOptions } from './queries';
import { formatCount, formatMoney, formatPercent, formatTime } from '@/lib/utils';

/**
 * 统计数据
 */
export interface StatsMetrics {
    input_token: number;
    // cached_token 与 cache_write_token 是输入中命中缓存和写入缓存的部分, 均为 input_token 的子集。
    cached_token: number;
    cache_write_token: number;
    output_token: number;
    input_cost: number;
    output_cost: number;
    wait_time: number;
    request_success: number;
    request_failed: number;
}

export interface StatsMetricsFormatted {
    input_token: ReturnType<typeof formatCount>;
    cached_token: ReturnType<typeof formatCount>;
    cache_write_token: ReturnType<typeof formatCount>;
    output_token: ReturnType<typeof formatCount>;
    input_cost: ReturnType<typeof formatMoney>;
    output_cost: ReturnType<typeof formatMoney>;
    wait_time: ReturnType<typeof formatTime>;
    request_success: ReturnType<typeof formatCount>;
    request_failed: ReturnType<typeof formatCount>;

    request_count: ReturnType<typeof formatCount>;
    total_token: ReturnType<typeof formatCount>;
    total_cost: ReturnType<typeof formatMoney>;
    // cache_rate 是缓存命中率百分比。原始值也一并保留: 渠道榜与模型榜要跨条目合并, 只能先按 raw 求和再重算。
    cache_rate: ReturnType<typeof formatPercent>;
}

// formatStatsMetrics 把一组累计统计格式化为界面展示字段，并补齐合计项。
// 渠道、渠道模型和渠道凭据各自维护独立计数，均按同一口径展示。
export function formatStatsMetrics(metrics: StatsMetrics): StatsMetricsFormatted {
    return {
        input_token: formatCount(metrics.input_token),
        cached_token: formatCount(metrics.cached_token),
        cache_write_token: formatCount(metrics.cache_write_token),
        output_token: formatCount(metrics.output_token),
        total_token: formatCount(metrics.input_token + metrics.output_token),
        input_cost: formatMoney(metrics.input_cost),
        output_cost: formatMoney(metrics.output_cost),
        total_cost: formatMoney(metrics.input_cost + metrics.output_cost),
        wait_time: formatTime(metrics.wait_time),
        request_success: formatCount(metrics.request_success),
        request_failed: formatCount(metrics.request_failed),
        request_count: formatCount(metrics.request_success + metrics.request_failed),
        cache_rate: formatPercent(cacheRate(metrics.input_token, metrics.cached_token)),
    };
}

// cacheRate 返回缓存命中率百分比，输入为空时为零。
// 命中 Token 是输入 Token 的子集，故分母取输入总数：未命中的部分同样计入输入。
export function cacheRate(inputTokens: number, cachedTokens: number): number {
    if (!inputTokens) return 0;
    return ((cachedTokens ?? 0) / inputTokens) * 100;
}

export interface StatsDaily extends StatsMetrics {
    date: string;
}
export interface StatsDailyResponse {
    max_request_count: number;
    items: StatsDaily[];
}
export interface StatsDailyFormatted extends StatsMetricsFormatted {
    date: string;
}
interface StatsDailyFormattedResponse {
    max_request_count: number;
    items: StatsDailyFormatted[];
}

export interface StatsTotal extends StatsMetrics {
    id: number;
}
type StatsTotalFormatted = StatsMetricsFormatted;

export interface StatsHourly extends StatsMetrics {
    hour: number;
    date: string;
}
interface StatsHourlyFormatted extends StatsMetricsFormatted {
    hour: number;
    date: string;
}
/**
 * API Key 统计数据
 */
export interface StatsAPIKey extends StatsMetrics {
    api_key_id: number;
}

export interface StatsAPIKeyFormatted extends StatsMetricsFormatted {
    api_key_id: number;
}

// statsDailyFormattedQueryOptions 统一首页每日统计查询、格式化和刷新策略。
const statsDailyFormattedQueryOptions = queryOptions({
    ...statsDailyQueryOptions,
    select: (data): StatsDailyFormattedResponse => ({
        max_request_count: data.max_request_count,
        items: data.items.map((item): StatsDailyFormatted => ({
            ...formatStatsMetrics(item),
            date: item.date,
        })),
    }),
    refetchInterval: 3600000, // 1 小时
    refetchOnMount: 'always',
});

/**
 * 获取每日统计数据 Hook
 */
export function useStatsDaily() {
    const query = useQuery(statsDailyFormattedQueryOptions);
    return {
        ...query,
        data: query.data?.items,
        maxRequestCount: query.data?.max_request_count ?? 0,
    };
}

// statsHourlyFormattedQueryOptions 统一首页每小时统计查询、格式化和刷新策略。
const statsHourlyFormattedQueryOptions = queryOptions({
    ...statsHourlyQueryOptions,
    select: (data) => data.map((item): StatsHourlyFormatted => ({
        ...formatStatsMetrics(item),
        hour: item.hour,
        date: item.date,
    })),
    refetchInterval: 10000,// 10 秒
    refetchOnMount: 'always',
});

/**
 * 获取每小时统计数据 Hook
 */
export function useStatsHourly() {
    return useQuery(statsHourlyFormattedQueryOptions);
}

// statsTotalFormattedQueryOptions 统一首页总统计查询、格式化和刷新策略。
const statsTotalFormattedQueryOptions = queryOptions({
    ...statsTotalQueryOptions,
    select: (data): StatsTotalFormatted => formatStatsMetrics(data),
    refetchInterval: 10000,// 10 秒
    refetchOnMount: 'always',
});

/**
 * 获取总统计数据 Hook
 */
export function useStatsTotal() {
    return useQuery(statsTotalFormattedQueryOptions);
}



/**
 * 获取 API Key 统计数据列表 Hook
 */
export function useStatsAPIKey() {
    return useQuery({
        queryKey: ['stats', 'apikey'],
        queryFn: () => apiRequest<StatsAPIKey[]>('/api/v1/stats/apikey'),
        select: (data) => data.map((item): StatsAPIKeyFormatted => ({
            ...formatStatsMetrics(item),
            api_key_id: item.api_key_id,
        })),
        refetchInterval: 30000,
        refetchOnMount: 'always',
    });
}

export interface StatsGroup extends StatsMetrics {
    group_id: number;
    group_name: string;
}

export interface StatsGroupFormatted {
    group_id: number;
    group_name: string;
    formatted: StatsMetricsFormatted;
}

// useGroupStatsByPeriod 取分组在指定天数窗口内的统计, days 为 0 表示累计。
// 按请求所属分组记一次, 不再把渠道模型总量按成员复制。
export function useGroupStatsByPeriod(days: number, enabled = true) {
    return useQuery({
        queryKey: ['stats', 'group', 'period', days],
        queryFn: () => apiRequest<StatsGroup[]>(`/api/v1/stats/group?days=${days}`),
        select: (data): StatsGroupFormatted[] => data.map((item) => ({
            group_id: item.group_id,
            group_name: item.group_name,
            formatted: formatStatsMetrics(item),
        })),
        enabled,
        refetchInterval: 30000,
        refetchOnMount: 'always',
    });
}
